File: MyFormsLibrary/FormWizard.vb

Partial Public Class FormWizard

    Protected _currentStep As Integer
    Protected _connection As System.Data.Common.DbConnection
    Delegate Function Esegui(ByVal parametri() As Object) As Boolean
    Private m_eseguiAzione As MyFormsLibrary.FormWizard.Esegui
    Private colorSelNode As Drawing.Color = Color.FromArgb(165, 203, 247)

    Public Property EseguiAzione() As Esegui
        Get
            Return m_eseguiAzione
        End Get
        Set(ByVal value As Esegui)
            m_eseguiAzione = value
        End Set
    End Property


    Public Sub New()
        MyBase.New()
        Me.InitializeComponent()

        Me._currentStep = 0
        ' AddHandler formc FormClosing, AddressOf Me.FormInsFormClosing

    End Sub

    'PIGLIACELLI: init vecchia *verificare su altri progetti se da rimuovere ()
    'Public Sub _init()

    '    'me.Mode = WorkMode.RunTime

    '    'format del treeview
    '    'dim myFont as New Font( FontFamily.GenericSansSerif, 9)

    '    'format tab controls

    '    For Each t As Windows.Forms.TreeNode In _treeView.Nodes
    '        t.BackColor = Drawing.SystemColors.Window
    '        t.ForeColor = Drawing.SystemColors.ControlText
    '        '		t.NodeFont = myFont
    '    Next
    '    _treeView.Nodes(0).BackColor = Drawing.Color.LightSkyBlue

    '    _ApplyGrant()

    'End Sub

    Public Overridable Function _init(ByRef connection As System.Data.Common.DbConnection, _
                                      ByVal titolo As String) As Boolean

        _tabControl.TabIndex = 0

        Me.Text = titolo
        Me.lblTitoloWizard.Text = titolo
        lblTitoloWizard.BackColor = colorSelNode


        If Not Me.Owner Is Nothing Then
            Me.Icon = Me.Owner.Icon
        End If
        Me._connection = connection

        For Each t As Windows.Forms.TreeNode In _treeView.Nodes
            t.BackColor = Drawing.SystemColors.Window
            t.ForeColor = Drawing.SystemColors.ControlText
            '		t.NodeFont = myFont
        Next
        _treeView.Nodes(0).BackColor = colorSelNode
        _currentStep = 0
        Me.btnPrevStep.Visible = False
        btnNextStep.Text = "Avanti"

        _ApplyGrant()

        Return True
    End Function



    Overridable Sub _GoToNextStep()
        Dim nextStep As Integer
        Dim MenuLeftIndex As Integer

        _currentStep += 1

        'navigazione normale ...
        If _currentStep < _treeView.Nodes.Count Then
            nextStep = _currentStep '+ 1
        Else
            nextStep = _currentStep - 1
        End If

        MenuLeftIndex = nextStep


        If _treeView.Nodes.Count > nextStep Then
            _treeView.Nodes(MenuLeftIndex - 1).BackColor = Drawing.SystemColors.Window
            _treeView.Nodes(MenuLeftIndex).BackColor = colorSelNode
        End If

       

        If _treeView.Nodes.Count > MenuLeftIndex Then
            label2.Text = _treeView.Nodes(MenuLeftIndex).Text
        End If


        'SE SALVATAGGIO-->!
        If _currentStep = _treeView.Nodes.Count Then

            If SaveAction() Then
                Windows.Forms.MessageBox.Show("Salvataggio avvenuto con successo", My.Application.Info.ProductName, Windows.Forms.MessageBoxButtons.OK, Windows.Forms.MessageBoxIcon.Information)
                'Dim formE As New MyFormsLibrary.FormMessageBox
                'formE._Init(MyFormsLibrary.FormMessageBox.Tipo.OK, "")
                Me.Close()
            Else
                'Dim formE As New MyFormsLibrary.FormMessageBox
                'formE._Init(MyFormsLibrary.FormMessageBox.Tipo.KO, "")
                Windows.Forms.MessageBox.Show("Salvataggio NON avvenuto con successo", My.Application.Info.ProductName, Windows.Forms.MessageBoxButtons.OK, Windows.Forms.MessageBoxIcon.Error)
            End If

        End If


        'ALLA FINE, SETTO IL CURRENT
        _tabControl.SelectedIndex = nextStep
        _currentStep = nextStep

    End Sub

    Overridable Sub _GoToBackStep()
        Dim backStep As Integer
        Dim MenuLeftIndex As Integer

        backStep = _currentStep - 1
        MenuLeftIndex = backStep

        If _treeView.Nodes.Count > _currentStep Then
            _treeView.Nodes(_currentStep).BackColor = Drawing.SystemColors.Window
            _treeView.Nodes(MenuLeftIndex).BackColor = Me.colorSelNode
        End If

        _tabControl.SelectedIndex = backStep
        _currentStep = backStep

        label2.Text = _treeView.Nodes(MenuLeftIndex).Text

    End Sub


    Sub FormInsFormClosing(ByVal sender As Object, ByVal e As Windows.Forms.FormClosingEventArgs)
        e.Cancel = (System.Windows.Forms.MessageBox.Show("La chiusura del wizard comporterà la perdita dei dati inseriti. Si desidera chiudere?", My.Application.Info.ProductName, System.Windows.Forms.MessageBoxButtons.OKCancel, System.Windows.Forms.MessageBoxIcon.Question) = Windows.Forms.DialogResult.Cancel)
    End Sub

    Public Overridable Sub BtnExitClick(ByVal sender As Object, ByVal e As EventArgs)
        Me.Close()
    End Sub


   

    Public Overridable Sub BtnNextStepClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnNextStep.Click
        _GoToNextStep()
    End Sub

    Public Overridable Sub BtnPrevStepClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnPrevStep.Click
        _GoToBackStep()
    End Sub

    Public Overridable Sub BtnLastStepClick(ByVal sender As Object, ByVal e As EventArgs)
        Windows.Forms.MessageBox.Show("Function not available", My.Application.Info.ProductName, Windows.Forms.MessageBoxButtons.OK, Windows.Forms.MessageBoxIcon.Error)
    End Sub

    Public Overridable Function SaveAction() As Boolean

        Windows.Forms.MessageBox.Show("Function not available", My.Application.Info.ProductName, Windows.Forms.MessageBoxButtons.OK, Windows.Forms.MessageBoxIcon.Error)
    End Function


    Public Overridable Sub _tabControlSelecting(ByVal sender As Object, ByVal e As TabControlCancelEventArgs) Handles _tabControl.Selecting

        '		If Mode <>  WorkMode.Designer andalso sender.GetType.Name =  "TabControl" Then
        '			e.Cancel = True
        '			Exit Sub
        '		End If

        Dim myTab As Windows.Forms.TabControl = DirectCast(sender, Windows.Forms.TabControl)
        'Dim lastIndex As Int16  = myTab.TabCount
        Select Case e.TabPageIndex
            Case 0
                Me.btnPrevStep.Visible = False
                btnNextStep.Text = "Avanti"
            Case myTab.TabCount - 1 'LAST STEP
                Me.btnPrevStep.Visible = True
                btnNextStep.Text = "Fine"
            Case Else
                Me.btnPrevStep.Visible = True
                btnNextStep.Text = "Avanti"
        End Select

        'directcast (sender, System.Windows.Forms.TabControl ).Selecting(sender)

        DirectCast(sender, System.Windows.Forms.TabControl).TabIndex = e.TabPageIndex
    End Sub

    Friend Overridable Function _ApplyGrant() As Boolean
        Return True
    End Function

End Class

File: MyFormsLibraryCSharp/FormWizard.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace MyFormsLibrary
{
    public partial class FormWizard : Form
    {

        protected  int _currentStep = 0;


        public FormWizard()
        {
            InitializeComponent();
        }


        public void _hideButtonIndietro()
        {
            btnIndietro.Visible = false;
        }


        public virtual bool _init(string title)
        {
            if (this.Owner != null && this.Owner.Icon != null)
            {
                this.Icon = this.Owner.Icon;
            }

            this.myTabControl1._initTabPageStatus();
            this.myTabControl1.Selecting += new TabControlCancelEventHandler(myTabControl_Selecting);

            if (myTabControl1.TabCount == 1)
            {
                btnAvanti.Text = "Fine";
              //  btnIndietro.Visible = false;
            }

            this.label1.Text = title;

            this.Text = this.GetType().Assembly.GetName().Name + " " + this.GetType().Assembly.GetName().Version.ToString() + " - " + title;
        

            btnIndietro.Visible = false;
            return true;
        }

        void myTabControl_Selecting(object sender, TabControlCancelEventArgs e)
        {
            if (e.TabPageIndex == myTabControl1.TabCount - 1)
            {
                btnAvanti.Text = "Fine";
            }
            else
            {
                btnAvanti.Text = "Avanti >>";
            }

           //base.OnSelecting(e);
        }

        private void btnAnnulla_Click(object sender, EventArgs e)
        {
            this.DialogResult = System.Windows.Forms.DialogResult.Cancel;
        }




        private void btnAvanti_Click(object sender, EventArgs e)
        {
            if (_currentStep == myTabControl1.TabPages.Count -1)
            {
                if (_checkBeforeSaveAction() == true)
                {
                    if (System.Windows.Forms.MessageBox.Show("Confermare l'operazione?", System.Windows.Forms.Application.ProductName, System.Windows.Forms.MessageBoxButtons.OKCancel, System.Windows.Forms.MessageBoxIcon.Question) == System.Windows.Forms.DialogResult.OK)
                    {
                        _saveAction();
                    }
                }
            }
            else
            {
                // navigazione normale
                _goToNextStep();
            }
        }



        public virtual bool _goToNextStep()
        {
            if (! _checkBeforeNextStep (_currentStep )) {
                return false;
            }
            _currentStep = _currentStep + 1;
            btnIndietro.Visible = true;
            myTabControl1.SelectedIndex = _currentStep;
            return true;
        }



        public void _goToStep(int stepNumber)
        {
            _currentStep = stepNumber;
            myTabControl1.SelectedIndex = _currentStep;
        }



        public virtual void _goToBackStep()
        {
            _currentStep = _currentStep - 1;
            myTabControl1.SelectedIndex = _currentStep;
        }

        public virtual bool _saveAction()
        {
            System.Windows.Forms.MessageBox.Show("Function not available", System.Windows.Forms.Application.ProductName, System.Windows.Forms.MessageBoxButtons.OK, System.Windows.Forms.MessageBoxIcon.Error);
            return false;
        }


        public virtual bool _checkBeforeSaveAction()
        {
            return true;
        }



        public virtual bool _checkBeforeNextStep(int stepIndex )
        {
            return true;
        }



        private void btnIndietro_Click(object sender, EventArgs e)
        {
            if (_currentStep == 0) {
                return;
            }

            if (_currentStep == 1)
            {
                btnIndietro.Visible = false;
            }
            _goToBackStep();
        }







    }
}
